Port "automation voice" uninstall logic to C++

jugglinmike opened the pull request:
This work builds on the gh-11, and as such, potential reviewers may want to hold off on offering feedback. The change set in gh-11 may evolve prior to being merged, and that may necessitate further modification to this patch. I'm submitting it as a "Draft Pull Request" to give some more visibility to the work, even before it's ready for formal review.

mzgoddard commented on src/makevoice/WindowsRegistry.cpp:
```diff
@@ -0,0 +1,137 @@
+#include <winreg.h>
+#include "WindowsRegistry.h"
+
+namespace WindowsRegistry
```
Should we nest this namespace, and other namespaces we create, in a namespace like AssistiveTechAutomation, ATAutomation, W3C, etc?

mzgoddard replied:
I deferred making this change for later if we want.

mzgoddard commented on src/makevoice/MakeVoice.cpp:
```diff
@@ -104,9 +179,64 @@ int wmain(int argc, __in_ecount(argc) WCHAR* argv[])
+
+int wmain(int argc, __in_ecount(argc) WCHAR* argv[])
+{
+    bool shouldInstall = parseArgs(argc, argv);
```
What do you think of returning the `DllAction` value instead of the boolean? Then compare it to determine the function to call?

mzgoddard reviewed (changes requested):
Looks good. 

I nitted some common things I noticed.

From the change it looks like you want to prefer `_T`, `tstring` and `TCHAR`. I commented on uses of `TEXT`, `WCHAR`, `L""` to standardize to that preference.

mzgoddard commented on src/makevoice/WindowsRegistry.cpp:
```diff
@@ -0,0 +1,137 @@
+
+        if (subKey.at(subKey.length() - 1) != TEXT('\\'))
+        {
+            subKey += TEXT('\\');
```
I moved the string concatenation into the loop as one expression `subKey + KEY_SEPARATOR + szName`.

mzgoddard commented on src/makevoice/MakeVoice.cpp:
```diff
@@ -104,9 +179,64 @@ int wmain(int argc, __in_ecount(argc) WCHAR* argv[])
+    return system("rmdir /Q /S \"" AUTOMATION_VOICE_HOME "\"") == 0 ? S_OK : E_FAIL;
+}
+
+int wmain(int argc, __in_ecount(argc) WCHAR* argv[])
```
This was a odd comment for me to leave. `wmain` doesn't support an ansi version, it'll always be `WCHAR`.

mzgoddard commented on src/makevoice/WindowsRegistry.cpp:
```diff
@@ -0,0 +1,137 @@
+            return lResult == ERROR_FILE_NOT_FOUND;
+        }
+
+        if (subKey.at(subKey.length() - 1) != TEXT('\\'))
```
I don't think this is every false. I replaced the if with single key path concatenation expression in the loop.

mzgoddard commented on src/makevoice/WindowsRegistry.cpp:
```diff
@@ -0,0 +1,137 @@
+            );
+        }
+
+        subKey.pop_back();
```
Replaced with `subKey + KEY_SEPARATOR + szName` expression in loop. Don't need to pop the `\\` anymore.

mzgoddard commented on src/makevoice/MakeVoice.cpp:
```diff
@@ -41,34 +53,97 @@ void copyFile(const char* sourceLocation, const char* destinationLocation)
+    {
+        if (_tcscmp(argv[i], L"/u") == 0)
+        {
+            return false;
```
Replaced with returned `DllAction` enum.

mzgoddard reviewed (approved):
I think this is ready now.

mzgoddard commented on src/makevoice/MakeVoice.cpp:
```diff
@@ -137,11 +142,11 @@ HRESULT install()
         hr = SpCreateNewTokenEx(
             SPCAT_VOICES,
-            TEXT(AUTOMATION_VOICE_ID),
+            L"" AUTOMATION_VOICE_ID,
```
I replaced `TEXT` uses here with `L""` preprocess concatenation here because as far as I can tell `SpCreateNewTokenEx` only takes `wchar_t` arguments.

mzgoddard commented:
I fixed an issue with the registry removal. It was trying to remove `SPCAT_VOICES "\\BocoupAutomationVoice"`. The registry entries we want to remove are `BocoupAutomationVoice` in `Tokens` in `SPCAT_VOICES`. So I've added a small commit to fix that.

jugglinmike commented on src/makevoice/MakeVoice.cpp:
```diff
@@ -12,11 +12,26 @@
+#define UTTERANCE_FOR_SETTING_DEFAULT_VOICE _T("Installing voice named " AUTOMATION_VOICE_NAME)
+
+// Registry path storing voice tokens.
+#define SPCAT_VOICE_TOKENS SPCAT_VOICES _T("\\Tokens")
```
```suggestion
#define SPCAT_VOICE_TOKENS _T(SPCAT_VOICES "\\Tokens")
```

@mzgoddard As written, this line took me a little research to understand, and I still don't know the character width of a literal like `"a" L"b"`. My confusion is certainly due to my unfamiliarity with macros, but I also think the version I'm proposing reads a little better (in addition to being consistent with the definition with precedes it).

jugglinmike replied:
On second thought, I'll propose this with a follow-up patch

jugglinmike replied:
Here it is: https://github.com/w3c/aria-at-automation-driver/pull/26

jugglinmike commented:
> I fixed an issue with the registry removal. It was trying to remove `SPCAT_VOICES "\\BocoupAutomationVoice"`. The registry entries we want to remove are `BocoupAutomationVoice` in `Tokens` in `SPCAT_VOICES`. So I've added a small commit to fix that.

I've confirmed the bug and your fix. Good catch!
